### PR TITLE
Add riak_ensemble_manager:subscribe and associated code  [JIRA: RIAK-1871]

### PR DESCRIPTION
**Pull request by nickelization**
This code allows you to set callbacks for ensemble state changes. It's currently needed for some data platform stuff, but potentially useful other places as well. Credit to Cuyler for writing the code - I'm submitting a PR on his behalf so we can get it into the develop branch.

### CODE REVIEW

**Comment by nickelization** [comment]
Okay, since I didn't actually write this code, I'm going to assume it's fine if I go ahead and review it. I've read over it, it looks good to me, and a bunch of people have been using it in development and testing already for the data platform work.

**Comment by nickelization** [comment]
:+1: 115241c

File: src/riak_ensemble_manager.erl [hunk 1]
```diff
@@ -369,6 +375,13 @@ handle_call({peer_info, Ensemble, Peer}, From, State) ->
             {noreply, State}
     end;
 
+handle_call({subscribe, Callback}, _, State=#state{subscribers=Subscribers}) ->
+    %% TODO: Change subscriber API to prevent multiple subscriptions with the same callback.
+    Subscribers2 = [Callback|Subscribers],
+    application:set_env(riak_ensemble, subscribers, Subscribers2),
```
Comment (hmmr) [inline src/riak_ensemble_manager.erl:381]: Passing a fun as an arg to application:set_env/1? Is that safe?

Comment (cuyler) [inline src/riak_ensemble_manager.erl:381]: In this case, this is the least frictionless way to keep some global state to persist even in the even of a crash of ensemble. Safety wise, it appears to be perfectly fine.

**Comment by JeetKunDoug** [comment]
@cuyler any thoughts on flushing out the API (unsubscribe?) and error handling (what if one of the funs fails)?

**Comment by JeetKunDoug** [comment]
@cuyler @nickelization Can you guys talk to error handling on subscriber callbacks in this case? Looks like we could have some issues if the callbacks fail for any reason? Also, unsubscribe would be good if we want to release a "complete API" for this.

**Comment by nickelization** [comment]
@JeetKunDoug I would agree it could certainly be improved, but since it was still useful as-is and fit the requirements of what was needed for the data platform, I went ahead and approved it without demanding additional enhancements. For now I'm still on the data platform team and probably would not be allowed to take time away from that to write more code for this feature. But it sounds like I'll be getting to switch to some strong consistency work in the near(-ish?) future, so at that point I'd be happy to improve error handling and add an unsubscribe call.

**Comment by JeetKunDoug** [comment]
Understand why it got put in for DP - just trying to figure out if we take this into a released Riak before the lack of error handling/unsubscribe functionality is taken care of - would rather wait until it's cleaned up before bringing this into Riak itself.

**Comment by nickelization** [comment]
@JeetKunDoug Well, to be fair, nothing in the base Riak code uses this subscribe functionality at all - it'll just be dead code in Riak proper. It might be half-baked, but lack of error handling or unsubscribe shouldn't compromise the stability of anything if nothing is subscribing for these events in the first place.

**Comment by nickelization** [comment]
Granted, someone might eventually see the code and decide to use it for something in Riak outside of the data platform, so I can totally agree with wanting to get it improved in the long term; just thinking about the short-term release ramifications, which is what I assume you're concerned about right now...?

**Comment by Basho-JIRA** [comment]
We've decided to remove this code from ensemble, as it is no longer needed. See https://github.com/basho/riak_ensemble/pull/104

Closing this JIRA issue accordingly. 

**_[posted via JIRA by Nick Marino]**_

**Pull request merged**